src/aeso_ast_infer_types.erl

```diff
@@ -862,17 +888,71 @@ infer1(Env, [{pragma, _, _} | Rest], Acc, Options) ->
     %% Pragmas are checked in check_modifiers
     infer1(Env, Rest, Acc, Options).
 
+check_implemented_interfaces([], _, _, _, _) ->
+    ok;
+check_implemented_interfaces([{contract_interface, _, IName, Extensions, Decls} | Interfaces], ConId, Impls, Acc, AllInterfaces) ->
+    case lists:member(name(IName), Acc) of
+        true ->
+            check_implemented_interfaces(Interfaces, ConId, Impls, Acc, AllInterfaces);
+        false ->
+            Unmatched = match_impls(Decls, ConId, name(IName), Impls),
+            NewInterfaces = [proplists:get_value(name(I), AllInterfaces) || I <- Extensions],
+            check_implemented_interfaces(Interfaces ++ NewInterfaces, ConId, Unmatched, [name(IName) | Acc], AllInterfaces)
+    end.
+
+match_impls([], _, _, Impls) ->
+    Impls;
+match_impls([{fun_decl, _, {id, _, FunName}, {fun_t, _, _, ArgsTypes, RetDecl}} | Decls], ConId, IName, Impls) ->
+    Match = fun({letfun, _, {id, _, FName}, Args, RetFun, _}) when FName == FunName ->
+                    length(ArgsTypes) == length(Args) andalso
+                    compare_types(RetDecl, RetFun) andalso
+                    lists:all(fun({T1, {typed, _, _, T2}}) -> compare_types(T1, T2) end,
+                              lists:zip(ArgsTypes, Args));
+               (_) -> false
+            end,
+    UnmatchedImpls = case lists:search(Match, Impls) of
+                         {value, V} ->
+                             lists:delete(V, Impls);
+                         false ->
+                             type_error({unimplemented_interface_function, ConId, IName, FunName}),
+                             Impls
+                     end,
+    match_impls(Decls, ConId, IName, UnmatchedImpls).
+
+
+-spec compare_types(T, T) -> boolean() when T :: utype() | [utype()].
+compare_types(Types1 = [_ | _], Types2 = [_ | _]) ->
+    length(Types1) == length(Types2) andalso
+    lists:all(fun({T1, T2}) -> compare_types(T1, T2) end, lists:zip(Types1, Types2));
```

Comment:
I don't think that `compare_types` is needed, that was meant to be a version of `unify` that does not give a type error (only returns true or false).

If I use `unify` here instead of `compare_types`, I will be getting 2 errors, one `cannot_unify` error and another `unimplemented_interface_function` where I only need the second one.

I can modify `unify` to not give a type error in one of three ways:
1. Pass a value to `When` and use that to indicate that we don't need to throw an error in this case (Probably a bad idea since the value `When` is only used for output)
2. Add a flag to `Env` (e.g. `unify_no_thorw`) and only throw if it's set to `true`
3. Add an argument to `unify` (e.g. `ShouldThrow`) and only throw if it's set to `true` (That would cause changes in multiple lines of the code in each definition of `unify`)

Reply:
Actually, I have faced the same issue while working on a different PR. I think that adding something like `unify_throws = true` to the record `env` could be the best idea here.

Reply:
What's wrong with unification errors here?

Reply:
I just want to check whether two types are the same or not, and move on if they are not. Unification errors will stop the compilation, and I don't want that.

Reply:
I have implemented that here 28bf14cc7cbda129dc7d4525e642b145698c1fea